### ISSUE TITLE
[RFC] Change terra-select to Data API from Children API

### ISSUE THREAD

**Issue by dev-cprice** [cerner/terra-core#2497]
# Feature Request

## Description

This is a companion RFC to https://github.com/cerner/terra-core/issues/2492.

We have been discussing using a data API instead of (or in addition to) the
children API for `terra-form-select`.

### Related
* https://github.com/cerner/terra-core/issues/2492
* https://github.com/cerner/terra-core/issues/2161
* https://github.com/cerner/terra-core/issues/2369

### Children API (current)

This is the current API for all `terra-form-select` variants. Simply pass a list
of `Select.Option` components as children to the `Select`.

<details>
<summary>Predefined Data Children API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PredefinedDataChildrenSelect() {
  // nothing special, hardcode a list of options here
  // NOTE: this is really rare in real world applications
  return (
    <Select placeholder="Select a color">
      <Select.Option value="blue" display="Blue" />
      <Select.Option value="green" display="Green" />
      <Select.Option value="purple" display="Purple" />
      <Select.Option value="red" display="Red" />
      <Select.Option value="violet" display="Violet" />
    </Select>
  );
}
```

</details>

<details>
<summary>Prop Driven Data Children API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PropDrivenDataChildrenSelect({ options }) {
  // map over list of options, transform shape if necessary
  // for this example, we use the shape { id: String, text: String },
  // instead of the expected { value: String, display: String }
  return (
    <Select placeholder="Select a color">
      {options.map(option => (
        <Select.Option
          key={option.id}
          value={option.id}
          display={option.text}
        />
      ))}
    </Select>
  );
}
```

</details>

<details>
<summary>Predefined Grouped Data Children API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PredefinedGroupedDataChildrenSelect() {
  // nothing special, hardcode a list of grouped options here
  // NOTE: this is really rare in real world applications
  return (
    <Select placeholder="Select a color">
      <Select.OptGroup label="Shade of blue">
        <Select.Option value="blue" display="Blue" />
        <Select.Option value="cyan" display="Cyan" />
        <Select.Option value="teal" display="Teal" />
        <Select.Option value="azul" display="Azul" />
        <Select.Option value="aqua" display="Aqua" />
      </Select.OptGroup>
      <Select.OptGroup label="Shades of green">
        <Select.Option value="green" display="Green" />
        <Select.Option value="forest" display="Forest Green" />
        <Select.Option value="dark" display="Dark Green" />
        <Select.Option value="neon" display="Neon Green" />
      </Select.OptGroup>
    </Select>
  );
}
```

</details>

<details>
<summary>Prop Driven Grouped Data Children API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PropDrivenGroupedDataChildrenSelect({ groups }) {
  // map over list of groups
  // for each group, map over list of options
  // no need to transform data for this example
  return (
    <Select placeholder="Select a color">
      {groups.map(group => (
        <Select.OptGroup label={group.label} key={group.label}>
          {group.options.map(option => (
            <Select.Option
              key={option.value}
              value={option.value}
              display={option.display}
            />
          ))}
        </Select.OptGroup>
      ))}
    </Select>
  );
}
```

</details>

<details>
<summary>All Children API Examples Usage & Render</summary>

```jsx
import React from 'react';
import { render } from 'react-dom';

function ChildrenApiExamples() {
  // this list could come from anywhere, or be any shape really
  // to illustrate data transformations (a common use case),
  // we use the shape { id: String, text: String }
  // instead of the expected { value: String, display: String }
  const options = [
    { id: 'blue', text: 'Blue' },
    { id: 'green', text: 'Green' },
    { id: 'purple', text: 'Purple' },
    { id: 'red', text: 'Red' },
    { id: 'violet', text: 'Violet' },
  ];

  const groupedOptions = [
    {
      label: 'Shades of blue',
      options: [
        { value: 'blue', display: 'Blue' },
        { value: 'cyan', display: 'Cyan' },
        { value: 'teal', display: 'Teal' },
        { value: 'azul', display: 'Azul' },
        { value: 'aqua', display: 'Aqua' },
      ],
    },
    {
      label: 'Shades of green',
      options: [
        { value: 'green', display: 'Green' },
        { value: 'forest', display: 'Forest Green' },
        { value: 'dark', display: 'Dark Green' },
        { value: 'neon', display: 'Neon Green' },
      ],
    },
  ];

  return (
    <div>
      <PredefinedDataChildrenSelect />
      <PropDrivenDataChildrenSelect options={options} />
      <PredefinedGroupedDataChildrenSelect />
      <PropDrivenGroupedDataChildrenSelect groups={groupedOptions} />
    </div>
  );
}

render(<ChildrenApiExamples />, document.getElementById('root'));
```

</details>

#### Pros

* Matches the native `<select>` API fairly closely, with a few changes
  * `<Select.Option>` uses `display` prop instead of `children` for `innerText`
* For hardcoded data, it's really easy to use
* Allows flexibility in using `<OptGroup>`s or `<Option>`s manually

#### Cons

* Doesn't handle dynamic data very well (more on this later)
* In real world examples, you need to manually map over data, applying
  transforms if necessary, and know that you need to render `<Select.Option>`
  and `<Select.OptGroup>` components here
* larger surface area in userland code === more surface area for bugs

### Data API

<details>
<summary>Predefined Data, Data API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PredefinedDataSelect() {
  // nothing special, hardcode a list of options here
  // can either define the data as a local, or just inline it as done below
  // NOTE: this is really rare in real world applications
  return (
    <Select
      placeholder="Select a color"
      data={[
        { value: 'blue', display: 'Blue' },
        { value: 'green', display: 'Green' },
        { value: 'purple', display: 'Purple' },
        { value: 'red', display: 'Red' },
        { value: 'violet', display: 'Violet' }
      ]}
    />
  );
}
```

</details>

<details>
<summary>Prop Driven Data, Data API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PropDrivenDataSelect({ options }) {
  // Thought: maybe we allow the consumer to pass an optional dataTransform
  // callback which allows the select to handle when the data items aren't
  // of the expected shape, like so:
  function dataTransform(option) {
    return { value: option.id, display: option.text };
  }

  // delegate the display of options to the Select itself
  return (
    <Select
      placeholder="Select a color"
      data={options}
      dataTransform={dataTransform}
    />
  );
}
```

</details>

<details>
<summary>Predefined Grouped Data, Data API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PredefinedGroupedDataSelect() {
  // nothing special, hardcode a list of grouped options here
  // NOTE: this is really rare in real world applications
  const groupedOptions = [
    {
      label: 'Shades of blue',
      options: [
        { value: 'blue', display: 'Blue' },
        { value: 'cyan', display: 'Cyan' },
        { value: 'teal', display: 'Teal' },
        { value: 'azul', display: 'Azul' },
        { value: 'aqua', display: 'Aqua' },
      ],
    },
    {
      label: 'Shades of green',
      options: [
        { value: 'green', display: 'Green' },
        { value: 'forest', display: 'Forest Green' },
        { value: 'dark', display: 'Dark Green' },
        { value: 'neon', display: 'Neon Green' },
      ],
    },
  ];

  // Note, this isn't necessarily a good idea here since we need to explore
  // a way to tell the select that the options are in fact grouped, another
  // prop is used here as a boolean, for example
  return (
    <Select
      placeholder="Select a color"
      data={groupedOptions}
      groupedOptions
    />
  );
}
```

</details>

<details>
<summary>Prop Driven Grouped Data, Data API Select Example</summary>

```jsx
import React from 'react';
import Select from 'terra-form-select';

function PropDrivenGroupedDataSelect({ groupedOptions }) {
  // Note, this isn't necessarily a good idea here since we need to explore
  // a way to tell the select that the options are in fact grouped, another
  // prop is used here as a boolean, for example
  return (
    <Select
      placeholder="Select a color"
      data={groupedOptions}
      groupedOptions
    />
  );
}
```

</details>

<details>
<summary>All Data API Examples Usage & Render</summary>

```jsx
import React from 'react';
import { render } from 'react-dom';

function DataApiExample() {
  // this list could come from anywhere, or be any shape really
  // to illustrate data transformations (a common use case),
  // we use the shape { id: String, text: String }
  // instead of the expected { value: String, display: String }
  const options = [
    { id: 'blue', text: 'Blue' },
    { id: 'green', text: 'Green' },
    { id: 'purple', text: 'Purple' },
    { id: 'red', text: 'Red' },
    { id: 'violet', text: 'Violet' },
  ];

  const groupedOptions = [
    {
      label: 'Shades of blue',
      options: [
        { value: 'blue', display: 'Blue' },
        { value: 'cyan', display: 'Cyan' },
        { value: 'teal', display: 'Teal' },
        { value: 'azul', display: 'Azul' },
        { value: 'aqua', display: 'Aqua' },
      ],
    },
    {
      label: 'Shades of green',
      options: [
        { value: 'green', display: 'Green' },
        { value: 'forest', display: 'Forest Green' },
        { value: 'dark', display: 'Dark Green' },
        { value: 'neon', display: 'Neon Green' },
      ],
    },
  ];

  return (
    <div>
      <PredefinedDataSelect />
      <PropDrivenDataSelect options={options} />
      <PredefinedGroupedDataSelect />
      <PropDrivenGroupedDataSelect groupedOptions={groupedOptions} />
    </div>
  );
}

render(<DataApiExample />, document.getElementById('root'));
```

</details>

#### Pros

* Simpler userland code === less surface area for bugs
* API is more focused and opinionated
  * Allows for easier access for optimizations on terra's end (e.g. windowing
    and async data)
* Delegates presentational responsibility of options to the `<Select>` itself
  * OCS doesn't allow you to roll custom options (yet lol)
* Still flexible enough to allow for grouped options, especially if we allow a
  first class prop for `dataTransform` and the interface shape is strictly
  defined and well documented

#### Cons

* Deviates from the native `<select>` API
* API is more focused and opinionated
* For hardcoded data, it's not as easy to use. But again, this is really rare in
  real world code.

### Asynchronous Data

The issues with the current `Select`'s implementation become clear when you are
using a multi-select variant. If you make a list of selections, they are simply
held in state [by value (id)](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_Frame.jsx#L390).

This is problematic because the selected options display as `<Tag>` components
which use the `value` (id, non-presentational text) to
[find the child with the correct value and get the display from](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/_SelectUtil.js#L85-L86) and use that
as the [child text](https://github.com/cerner/terra-core/blob/master/packages/terra-form-select/src/Select.jsx#L162)
(Related, see: https://github.com/cerner/terra-core/issues/2369)

Could we simply cache the entire selected object?
Sure. This would allow us to not have to search through the children for a child
with a specified value to get the presentational text. Actually, that's what I'm
proposing here. We could do the following:

<details>
<summary>Example code</summary>

```jsx
this.state {
  selected: []
};

function addSelection(option) {
  // this is just for demonstrating the expected shape
  const newSelection = {
    value: option.props.value,
    display: option.props.display,
  };

  this.setState(state => ({
    selected: [...state.selected, newSelection];
  }))
}

function removeLastSelected() {
  this.setState(state => ({
    selected: state.selected.slice(0, state.selected.length - 1),
  }));
}

function removeSelectedByValue(value) {
  this.setState(state => ({
    selected: state.selected.filter(selection => selection.value !== value),
  }));
}
```

</details>

This is actually a slight tangent, but is still related to the current children
implementation since we
***search through the children to find the display of the option with the
matching value***.

### Allowing both APIs at the same time

We ***could*** do this as well, but I'm cautious here since I think we should err on the side of simpler implementation and an opinionated API. It is also possible to limit the children API to only the default variant since it shouldn't contain dynamic data or long lists, however, I question the overall value of doing this.

---

Do you think this is a feature you would use? Could it simplify your userland code? Do you "Thanks, I hate it"? Comments welcome!

**Comment by emilyrohrbough** [comment]
If I am reading this correctly, is this API proposal for removing the need for consumers to do their own data looping?

As a user less familiar with select, it seems like moving to an object/array oriented prop type isn't much different than directly supplying props to a component since props are just an object. Moving to this type of syntax would requires strict Object structures for select to correctly loop and create options/opt groups, which might be more confusing/more error pron. Also if we add in transform props so users can control how their data might be generated makes me think users should just manager their own select data.

**Comment by dev-cprice** [comment]
@emilyrohrbough It's more than that. That's simply one possible side effect.

This suggested API is similar to [react-select](https://react-select.com/home), a popular `<select>` library implementation in React. An example of grouped data is [here](https://codesandbox.io/s/hhgf4?module=/example.js), with the shape looking like:

![image](https://user-images.githubusercontent.com/34115417/60622899-fc079c80-9da6-11e9-8fff-bbabc4bc89cb.png)

As far as the transform prop goes, that's not necessarily what I'm suggesting we do, but it is an option. The example from before was just a hypothetical addition. The user could just as well transform the data before they pass it to the `<Select>`, like so:

```jsx
function Example({ options }) {
  const data = options.map(option => ({ value: option.id, display: option.text }));

  return (
    <Select data={data} />
  );
}
```

That's already being done today. 

---

That being said, let me ask you a couple of questions:

For the current `terra-select` implementation:
1. How would you expect to handle asynchronous data? As in the following example:

```jsx
import React, { useState, useCallback } from 'react',
import useFetch from './use-fetch'; // pretend this exists

function AsynchronousDataExample() {
  // simple api, the values only change when doFetch(url) is called and the promise resolves
  const { data, isLoading, isError, doFetch } = useFetch();

  const handleSearch = useCallback((searchValue) => {
    // perform a fetch with the given baseUrl and searchValue as a query param
    doFetch(`http://some.api.here?q=${searchValue}`);
  }, []);

  return (
    <Select variant="multiple" onSearch={handleSearch}>
      {data.map(({ value, display }) => <Select.Option key={value} value={value} display={display} />)}
    </Select>
  );
```

or, put a little more simply, the options change out from under the select when a search is performed on the select's input. How do you maintain selections when they map only to ***options that are currently children***. (My suggestion under *Additional Information > Asynchronous Data* in the OP fixes this, but alludes to a bigger problem, hence this RFC).

This breaks the current implementation and is a valid use case: filtering/limiting option list size to a much smaller subset of a larger set (e.g. 100 vs 10000).

2. How would you expect to handle windowing (for performance gains)?

3. Is there a case where we custom `<Foo.Option>`s?

4. Is it not strange to have to know that you need to do either:

```jsx
import Select from 'terra-form-select';

<Select>
  <Select.Option ... /> // <= this part, Select.Option
</Select>
```

or the also proposed:

```jsx
import Select, { Option } from 'terra-form-select';
```
?

**Comment by dev-cprice** [comment]
One other thing that I'm alluding to is that I think there needs to be a new component that handles the search for and selection of a smaller subset from a large set first class, like this: https://github.com/cerner/terra-core/issues/1590

**Comment by emilyrohrbough** [comment]
Like I said, I'm not familiar and I def don't have the answers for this, so maybe my input is irrelevant. 😄 just my initial thoughts so bare with me (also, I am having a hard time following your examples given the length of your comments... maybe using html collapsable sections or inline-diffs might break this down a little):

1. Regardless of how options are rendered, when a search is performed, and previous options are "blow away", yet selected, won't the multi-select variant need to manage an internal `selection` list with props (or data object) of information anyways. Or how would the user be able to click away and then re-click on the select input and see the list of selected options?

I guess with the proposed Data API, are you expecting select to keep a continuous list of all data options that it was ever given?

4. Why is this strange? A majority of our current API is like this.

**Comment by dev-cprice** [comment]
Ope, sorry @emilyrohrbough. 🙃I'll break my previous replies up some.

---

1. So, you're correct. The multi-select does need to maintain an internal selection list, and it does. However, it only keeps the `value`s, instead of both the `value` and `display`. This is an internal implementation detail of the `terra-select` that is lacking. I think this will illustrate the situation a little better: https://github.com/dev-cprice/terra-select-dynamic-data

<details><summary>Here's a gif of it in action:</summary>

![2019-07-03 19 09 33](https://user-images.githubusercontent.com/34115417/60631665-2c126800-9dc6-11e9-83d6-52e5dac1f23e.gif)

</details>

I'm not suggesting that we keep a continuous list of all data options ever given, although _we could_ cache values in a hash based on the user input (react-select has a prop to do this too):

<details><summary>Example simple cache:</summary>

```js
const searchValue = 'foobarbaz';
const data = [/* imagine some data here */];
const cache = {
  [searchValue]: data,
};
```

</details>

I'm suggesting we need to be keeping a list of ***all selections***.

---

4. It's strange in the sense that it *feels leaky*. I'll admit that this is a nit, but it also leads to a lot of boilerplate code. I can't tell you how many times I've written the following 😂:

<details><summary>Code snippet</summary>

```jsx
<Select>
  {data.map({ value, display }) => <Select.Option key={value} value={value} display={display} />)}
</Select>
```

</details>

**Comment by emilyrohrbough** [comment]
> I'm suggesting we need to be keeping a list of all selections.

Ah, yeah, I would have expected this type of behavior. I assume we are doing the same with selectable list or table

**Comment by yuderekyu** [comment]
> Allowing both APIs at the same time

I am fine with this if we document the data pattern as preferred for asynchronous purposes and create a site example. I favor keeping the children API because of aforementioned reasons by others.

**Comment by yuderekyu** [comment]
> Is there a use case where we allow or use custom <Foo.Option>s?

We have some discussion about that here - https://github.com/cerner/terra-core/issues/2254. Seems like we should not open up options to be custom. Instead, if the use case arises, a new component should be created to handle it.

**Comment by bjankord** [comment]
I'm in favor of switching select and its related variants to a data prop API approach instead of using the children API for options/opt-groups. 

In terms of pros, there is one driving factor that persuades me to switch fully to a data prop API instead of supporting both a children based and data prop driven API, and that is less code usually means less bugs. 

The select already has a high level of complexity due to the nature of the functionality it provides and accessibility concerns it needs to address. While I like the idea of reaching a compromise and providing both a child based API (for passivity) and a data prop API, I have concerns that by supporting both, we will continue to have a high level of complexity within the component that will make working on it, (adding enhancements, fixing bugs in it), a non-trivial task for most contributors.

I'd like to hear thoughts from @mjhenkes @ryanthemanuel @tbiethman @StephenEsser and other consumers that use the terra-select component.

**Comment by StephenEsser** [comment]
For asynchronous operations I think users would favor the data driven API, but I would like to see an analysis of how and why the data API reduces the overall amount of code and complexity within the Select as is alluded to with this RFC.

Many of the arguments presented for the data API can be resolved in the same way, or in a similar way, with a children API. The arguments themselves actually appear to add more code and complexity. This RFC itself presents the potential of additional props for data transforms and managing the opt groups. `dataTransform` and `groupedOptions`. The usage of groups in general appears ambiguous and undecided with the data API whereas this is a non-issue with the children API.

> In real world examples, you need to manually map over data, applying
transforms if necessary, and know that you need to render <Select.Option>
and <Select.OptGroup> components here

The data transform and mapping is an important discussion point. The data returned from a service will likely not match the API necessary for the select options. So the data will need to be massaged into the correct format in both the data API and the children API. The mapping that is repeatedly marked as a pain point with the existing children API is also present in the data driven API. In the scenario where the `dataTransform` isn't a part of the API the users will still need to map and transform the data returned from a service into correctly formatted options. This would be a requirement for either API, the data driven API does not solve the mapping. It only moves the logic into the additional `dataTransform` (hypothetically this gets added) prop which would make the prop type for options unknown and unstructured. The user could pass any shape and transform the data appropriately for the options. This allows a high level of freedom for the user but also means the shape of the options is unknown and cannot be reliably interpreted.

> Doesn't handle dynamic data very well (more on this later)

> So, you're correct. The multi-select does need to maintain an internal selection list, and it does. However, it only keeps the values, instead of both the value and display. This is an internal implementation detail of the terra-select that is lacking. I think this will illustrate the situation a little better:

Caching the selected options for dynamic loading would resolve the behavior of the displays reverting back to their base values for both the children and data API. This is not inherently an issue with the children API, this is an issue that is present in both the data API and the children API. A cache of selection options would resolve the issue in either API.

> larger surface area in userland code === more surface area for bugs

I don't think enough data has been collected to support this. The data mapping / transforms will be required for both API proposals. The only real difference is where and what objects the data is mapped into. `<Select.Options>` vs `{}`.

> Is there a use case where we allow or use custom <Foo.Option>s?

Many teams have reached out with this question wanting to implement their own custom options. Although we are not directly supporting this, many teams are wanting this functionality. Our decision thus far has been not to allow custom options, largely for accessibility reason, but the want and need are present.

> Is it not strange to have to know that you need to do either:

No, this is not strange. This is the behavior many React components follow. It's also a very similar behavior you'd expect with native html. Most engineers that have prior experience with a Select would likely be familiar with the native select, which uses a children API, or Select2 which also uses the children API.  

Fundamentally the children API is just an object. This object could theoretically be treated the same as the `options` prop within the data API. I haven't seen any use-case presented that the data API resolves that cannot also be resolved in a near identical way with the children API.

There is also a really high tax that hasn't been addressed with moving to the data API. This change will impact every consumer of the Select to transform their APIs. The changes required by splitting the packages are intended to be small initially, this API change will be a lot more impactful.

Many of the pros and cons mentioned above apply to both APIs and would be resolved in a similar way.

All of that said. I think we should continue to discuss and consider the data API, fundamentally I think the decision should come down to "which API is easier to use for the 90% use-case".

**Comment by bjankord** [comment]
The non-passive change from a children API to a data API is a concern to me. I'm interested to hear from more consumers about what their thoughts are on this proposed API change as they will be the ones using it / interfacing with it.

The following users had logged issues to terra-core about terra-select. I'm curious if any of them have feedback on this RFC.

@johnaudas, @bbaraban, @kschuste, @shinepd, @prmeznar, @PolymorphedSquirrel, @dv297, @arifakram05, @BrandonAshbee, @kc4181, @WhiteLight1107, @nagrawal3, @ChaseSinclair, @dr054448, @PseudoFaux, @hboege, @mayurikini, @doogthemushroom, @dnguyenchevy, @madelineisabelle

**Comment by prmeznar** [comment]
As @StephenEsser outlined above, I don't really see many pros/cons from one implementation to the other themselves, except that one already exists. I'd want a pretty clear understanding of why the new approach was better before being asked to handle non-passive changes, and I don't have that understanding (yet, maybe).

Since we're talking about it, I'd like +1 to some sort of cache of selection options (we've had to implement some weirdness due to our unusual workflow of default values coming from one place and the actual options coming from another). But I'd rather have that change be standalone.

**Comment by dev-cprice** [comment]
## Caching Selections

For the caching of selections, I should've been more clear that this was an aside and tangentially related. Not that it's a solution that one API is immune to or not.

---

## On Passivity

I do agree that this would not be a passive change (that's what major versions are for 😉) and it's smart to limit churn here. I don't think that the refactor would be much on users end (in fact, they'd probably end up with less code), but even a minor refactor is a huge deal.

***Lots of value gained (maintained?) through passivity.***

---

## On Async Data

Being as this is a sister discussion to #2492, I'm gonna go into a slight tangent, but will discuss the main value of the data API, which comes from an asynchronous data approach.

### Defining Async Data

By asynchonous data, I mean that the select's options are being dynamically changed at runtime due to user input, ***not*** asynchronous load a list of options (which are unchanging after load) and then render the select.

### Applicable To..

Asynchronous is really only applicable for a searchable variant (the current variants: `multiple` and `search`). This variant has an input built into the select (yes, so does tags and combobox, but those are simply editable multiple and search variants respectively) that currently allows you to filter down the list of options to a subset dependent on the input value.

### Painting the Picture

* want to select 1 or more options from our list
* large lists of data, e.g. >>= 1000(s) of elements

As noted in #2161 and explored [here](https://github.com/dev-cprice/terra-select-perf), rendering large lists of data in the select options is problematic: it hangs the UI thread. 

So, let's make it async.

### (A) Solution

Using a children API here seems... odd and problematic. By delegating the rendering of children here to the consumer, it means the handling logic is on them. This means a lot of boilerplate will be needed in userspace code for:

* when the input changes you'll need to refire async calls
* throttling/debouncing
* search value thresholds (e.g. entering X characters before a search fires)

Of course, this all can be done with select now (mostly), but it requires a large amount of code to do so (part of which includes the caching of selections for multiselect). These things seem to be first class concerns of an async data based component. 

Instead of children, a common approach to solve this, is to fetch from a service subsets of options and render those, as opposed to the entire list. To incorporate such a workflow into `terra-select`, a common approach from other libraries is to use a `loadOptions` prop that returns a `Promise` which resolves to a list of data. This keeps it async loading implementation agnostic (e.g. `fetch`, `axios`, `$.ajax`, `xhr`, etc), e.g.:

```js
// in loadOptionsWith*() methods, searchValue comes from the Select's internal input

function loadOptionsWithFetch(searchValue) {
  return fetch(`some.api/here?q=${searchValue}`).then(response => response.json());
}

function loadOptionsWithAxios(searchValue) {
  return axios.get(`some.api/here?q=${searchValue}`);
}

<Select loadOptions={loadOptionsWithFetch} />
// OR
<Select loadOptions={loadOptionsWithAxios} />
// ETC
```

In this scenario, we cannot expect that the data returned is of the shape we expect. And rightfully so! The presentational layer should not dictate the backend; it's always the other way around since UI is so cheap/easy to change. So, what do we do? Well, we have two options:

1. in the `loadOptions` call, simply massage the data in your `Promise`, e.g. `data.map(e => ...)`
2. provide a first class way to transform each item as we map over them in the select when rendering the `<Option>`s, e.g.

```js
// this would be called on each individual list item when rendering options
function transformData(data) {
  const { id, text } = data;
  
  return {
    value: id,
    display: text,
  };
}
```

either way, we need to have a strict definition of the expected shape of each option.

What about the listed concerns above (throttling/debouncing, handling when to fire async calls, search thresholds, etc)? Well, they're better handled as internal implementation details in the select, driven by props of course. This is much easier to achieve with the `loadOptions` or `data` API just described.

---

That being said, would it be nice to have a shared/similar API between this async search behavior and the current select variants? Sure. However, this isn't a deal breaker and not a hill worth dying on.

I think the main goals behind #2492 is ***simplication*** and ***maintainability***. I think we go with either data or children, but not both for the sake of those goals. It appears the consensus so far is that there isn't enough value in converting the other variants over to the `data` API, which is fine 🙂.

**Comment by prmeznar** [comment]
Sounds like a blending of [terra-search-field](https://engineering.cerner.com/terra-ui/#/components/terra-search-field/search-field/search-field) and the multiple variants of select. I'm out of my depth here just as a consumer of select, but I wonder how blended those concepts are intended to be, and if starting from select, search, or a separate component entirely would be the correct way to blend them.

**Comment by dev-cprice** [comment]
@prmeznar in a way, it is. There's cross-over in some features here, sure.

search field is a way to roll your own search workflow, but it's a very generic interpretation. The workflow described above is primarily a selection one, with the optimization of lazily loading subsets of a much larger list of data.

There's been talks before about making a first class component for this exact use case, but it does not exist today.

**Comment by PolymorphedSquirrel** [comment]
While I'm not opposed to change in the API, I'm having a hard time seeing the immediate value of this versus the burden that it could potentially put on the consumers.  While there would be some shifting of complexity with the data API, it must exist at some level.  In my application, we use the terra-form-select in some very unorthodox ways to be able to satisfy our requirements.  I would be very hesitant to take on a new version without some very thorough vetting that the new design would fully support our usage.  

The children pattern, as it stands now, is allowing us a relatively simple and maintainable way to allow us to coerce the select into doing some very particular operations, and allowing us to override some of the undesired default behaviors for our use cases.  If any of this were to break, a core functionality of our application would become unstable and require investigation into new ways to elicit the desired functionality while still maintaining the feel and consistency with the rest of the framework.

Given the dependencies brought along by terra components (which result in multiple terra versions being pulled into your bundle when some versions stay behind), it would be unpleasant for us to willfully fall behind until finding capacity to come up with a new implementation.  I'm pretty risk-averse on this matter.  I would rather stay with the children API until an alternative could be completely proven out and shown to be as passive as possible while still maintaining the ability to short-circuit undesired behaviors for consumers with valid use-cases for doing so.

While I can respect the desire for better supporting async operations, I'd almost rather see that as a further specialized component (in terms of say typing in a partial name, pulling back a potential of users matching the input text, allowing a selection, maintaining that selected name in the options, searching another and adding a second).  I would expect that the consumer would still be responsible for guaranteeing that the selected entries be tracked in their state and be able to populate any necessary data for forcing the rendering of that option in a selected status to the dropdown while continuing any service calls and changes in the current list of reflected based on them.

Sorry, I'm a bit late to the party (and a bit exhausted), but hopefully still coherent enough to try to get these thoughts across.

**Comment by bjankord** [comment]
I re-read through this and some things that stood out to me were:
* The desire that people want to maintain passivity.
* A core issue this RFC is aiming to address is managing the select's options that are being dynamically changed at runtime due to user input in the multiple and search field. 
* Potential for a new component to handle this functionality.
* A desire to create some sort of cache of selection options.

Some options that I see that could address these:
* Applying the data API only to the multiple and search variants. This would be non-passive but would minimize the impact of the change to consumers instead of having the entire API switch over for the select component.

* Creating a new component specifically handling the search/filter use case that uses a data prop for API from the beginning.

I'd like to gather more feedback from current consumers to understand their thoughts on the RFC.

I also think it would be helpful to create a reduced test case to illustrate the problems you've faced @dev-cprice  when using the children API for managing the select's options that are being dynamically changed at runtime due to user input in either the multiple or search variant.

**Comment by calebmeyer** [comment]
Sorry if I'm resurrecting this, but I'm one of those users (outside of the terra team) who wishes (very strongly) that there were some option for selecting that supported:
- search
- maybe multi-select
- ajax results

Prior to moving off of blue steel, we just used Select2 for this, which handles all of these well, if not perfectly. Now we've got a bunch of bad solutions implemented throughout our code:
- handle caching selected values separately
- handle ajax calls (including both search and initial options) separately
- handle indicating loading to the user via hacks (using a span with an icon-spinner for the display value)

In every place we've implemented it, it's done a little differently. Having a component which handles all this for us (except generating the options at the server) would be incredible.
